polar-core/src/kb.rs

```diff
@@ -592,6 +586,35 @@ impl KnowledgeBase {
         let error: PolarError = error.into();
         error.set_context(source.as_ref(), Some(term))
     }
+
+    pub fn rewrite_implications(&mut self) -> PolarResult<()> {
+        let mut errors = vec![];
+
+        errors.append(&mut super::sugar::check_all_relation_types_have_been_registered(self));
+
+        let mut rules = vec![];
+        for (namespace, implications) in &self.namespaces.implications {
```

Comment:
Nit: should this drain implications so that we can never rewrite twice?

Reply:
That'd be a mutable borrow (for the `.drain()`) and an immutable borrow inside the loop. There's probably a better way to write this, but for now I don't think it's worth investing any more time. I added an unconditional `.clear()` of the namespaces as soon as we're done needing them: 4069e80